packages/zwave-js/src/lib/driver/Driver.ts

```diff
@@ -2412,6 +2413,53 @@ ${handlers.length} left`,
 				if (!msg.command.moreUpdatesFollow) {
 					nodeSessions.supervision.delete(msg.command.sessionId);
 				}
+			} else if (
+				msg.command.encapsulatingCC instanceof SupervisionCCGet
```

Comment:
Agreed - also see my comment here: https://github.com/zwave-js/node-zwave-js/issues/3201#issuecomment-904993646

We're going to need to remember how a command was encapsulated and use that flag to check whether we need to respond to SupervisionGet.

Reply:
Correct me if I'm wrong, but wouldn't using [`msg.command.isEncapsulatedWith`](https://github.com/zwave-js/node-zwave-js/blob/master/packages/zwave-js/src/lib/commandclass/CommandClass.ts#L995) instead of just `msg.command.encapsulatingCC` work for this? Since that method loops through each CC's `encapsulatingCC` so if `SupervisionCCGet` was an encapsulating CC somewhere in the chain that would find it?

I wasn't ignoring your original suggestion btw @AlCalzone - I was just trying to get the thing to work in some fashion before moving onto the finer details to make it robust :)

Reply:
@IvanBrazza `isEncapsulatedWith` only works until the command was unwrapped with `unwrapCommands`, which happens before the request handlers. During that process, the encapsulating CCs are discarded and only the innermost one is preserved to be handled.

Reply:
Ah I see that now - will look to implement something to remember the encapsulation chain during unwrapping.

Reply:
I have implemented this in the form of a new `encapsulatingCC` array on the `Message` which the `unrwrapCommands` class populates. It also saves the encapsulating `SupervisionCCGet` CC into a `supervisionEncapsulation` property which is then used by the `handleRequest` method to determine if the command was encapsulated with `SupervisionCCGet` at some point. How does that look to you guys?

Reply:
See https://github.com/zwave-js/node-zwave-js/pull/3240#pullrequestreview-746628963

Reply:
So I was taking a closer look at the existing `isEncapsulatedWith` method, and did a little testing and as far as I can see it does preserve the whole chain because you can follow the chain back up from the innermost CC.

I added the following code inside the existing `} else if (msg.supervisionEncapsulation) {` that was added via this PR:

```typescript
...
			} else if (msg.supervisionEncapsulation) {
				let encapstr = "";
				let encap = msg.command.encapsulatingCC;
				while (encap) {
					encapstr = `${encapstr}-${encap.ccName}`;
					encap = encap.encapsulatingCC;
				}
				const encapWithS2 = msg.command.isEncapsulatedWith(
					CommandClasses["Security 2"],
                                        Security2Command.MessageEncapsulation,
				);

				const encapsWithSupervisionGet = msg.command.isEncapsulatedWith(
					CommandClasses.Supervision,
					SupervisionCommand.Get,
				);

				this.controllerLog.logNode(
					msg.command.nodeId,
					`encapstr: '${encapstr}', encapWithS2: ${encapWithS2}, encapsWithSupervisionGet: ${encapsWithSupervisionGet}`,
				);

				if (
					msg.command instanceof InvalidCC &&
...
```

Which results in a log output of:

```
2021-09-05T13:11:20.738Z DRIVER « [Node 047] [REQ] [BridgeApplicationCommand]
                                  └─[Security2CCMessageEncapsulation]
                                    │ sequence number: 165
                                    │ extensions:
                                    │ · type: SPAN
                                    │   sender EI: 0xd4d3c35826cc3a7c0c07f55c2dc96db7
                                    └─[SupervisionCCGet]
                                      │ session id:      3
                                      │ request updates: false
                                      └─[NotificationCCReport]
                                          notification type:   Home Security
                                          notification status: 255
                                          notification state:  Motion detection
2021-09-05T13:11:20.742Z CNTRLR   [Node 047] encapstr: '-Supervision-Security 2', encapWithS2: true, encapsWith
                                  SupervisionGet: true
```

So you're able to do things like `msg.command.isEncapsulatedWith(CommandClasses["Security 2"]);` and the method will follow it all the way up the chain.

Am I missing something obvious here?

Reply:
I'll check it out. If it works that way, it wasn't intended but that should still do it 🤣

Reply:
Ok, you're right. The fact that we track both the encapsulated and encapsulating CC makes this possible.
Then let's get rid of the added tracking code and just use
```
msg.command.isEncapsulatedWith(CommandClasses.Supervision, SupervisionCommand.Get)
```
to check if we need to respond to a Supervision Get after handling the command.

Reply:
Awesome 😄 I have now updated that - I believe that resolves all outstanding comments!